File: src/main/java/com/udalmik/algs/interviewbit/stackqueue/MinStack.java
package com.udalmik.algs.interviewbit.stackqueue;

/*
 * Stack with constant time for "Get Min" operation
 */
public class MinStack {

    private Node head;

    public void push(int x) {
        if (isEmpty()) {
            head = new Node(x, null, x);
        } else {
            Node node = new Node(x, head, Math.min(head.minValue, x));
            head = node;
        }
    }

    public void pop() {
        if (!isEmpty()) {
            head = head.next;
        }
    }

    public int top() {
        if (isEmpty()) {
            return -1;
        } else {
            return head.value;
        }
    }

    public boolean isEmpty() {
        return head == null;
    }

    public int getMin() {
        if (isEmpty()) {
            return -1;
        } else {
            return head.minValue;
        }
    }
}

class Node {
    int value;
    Node next;
    int minValue;

    public Node(int value, Node next, int minValue) {
        this.value = value;
        this.next = next;
        this.minValue = minValue;
    }

}

File: js/leetcode/stack/min-stack.js
class MinStack {
    _data = [];
    _min = [];
    push(x) {
        this._data.push(x);
        if (this._min.length === 0) {
            this._min.push(x);
        } else {
            this._min.push(Math.min(this._min[this._min.length - 1], x));
        }
    }
    top() {
        return this._data[this._data.length - 1];
    }
    pop() {
        this._min.pop();
        return this._data.pop();
    }
    getMin() {
        return this._min[this._min.length - 1];
    }
}

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(x)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */